In [73]:
import pandas as pd
import numpy as np
import os
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt

In [74]:
df_final_demo=pd.read_csv('initial_datasets/df_final_demo.txt')
df_final_demo.head()


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1    2304905             7.0              94.0      58.0     U        2.0   
2    1439522             5.0              64.0      32.0     U        2.0   
3    1562045            16.0             198.0      49.0     M        2.0   
4    5126305            12.0             145.0      33.0     F        2.0   

         bal  calls_6_mnth  logons_6_mnth  
0   45105.30           6.0            9.0  
1  110860.30           6.0            9.0  
2   52467.79           6.0            9.0  
3   67454.65           3.0            6.0  
4  103671.75           0.0            3.0

In [75]:
print(df_final_demo.shape)
print(df_final_demo.dtypes)
print(f'\nnull values \n{df_final_demo.isnull().sum()}')

(70609, 9)
client_id             int64
clnt_tenure_yr      float64
clnt_tenure_mnth    float64
clnt_age            float64
gendr                object
num_accts           float64
bal                 float64
calls_6_mnth        float64
logons_6_mnth       float64
dtype: object

null values 
client_id            0
clnt_tenure_yr      14
clnt_tenure_mnth    14
clnt_age            15
gendr               14
num_accts           14
bal                 14
calls_6_mnth        14
logons_6_mnth       14
dtype: int64


In [76]:
df_data1 = pd.read_csv("initial_datasets/df_final_web_data_pt_1.txt")
df_data1.head()

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

In [77]:
print(df_data1.shape)
print(df_data1.dtypes)
print(f'\nnull values \n{df_data1.isnull().sum()}')
print(f'\nduplicated values\n{df_data1.duplicated().sum()}')
print(f'\nduplicated rows\n{df_data1[df_data1.duplicated()]}')

(343141, 5)
client_id        int64
visitor_id      object
visit_id        object
process_step    object
date_time       object
dtype: object

null values 
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

duplicated values
2095

duplicated rows
        client_id             visitor_id                      visit_id  \
365       9358626  148475529_63668223960  223297395_36250329195_832161   
368       5376894  622509328_44408201172  688984457_43441834354_912755   
383        438567  923318467_50708963627  330543236_29863358529_771432   
785       1423857  716830992_81930238382   568819166_80596727019_75889   
1816       372523  730241475_25771921122   25331354_74479847014_854987   
...           ...                    ...                           ...   
342621    1853345  831456815_77348510380  596026060_87178689141_910423   
342677    1251269  791822615_78146852845   763845944_55757914723_70888   
342679    1251269  791822615_781468

In [78]:
df_data1[df_data1.duplicated(subset=['visitor_id', 'visit_id'], keep=False)]

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
343136    2443347  465784886_73090545671  136329900_10529659391_316129   
343137    2443347  465784886_73090545671  136329900_10529659391_316129   
343138    2443347  465784886_73090545671  136329900_10529659391_316129   
343139    2443347  465784886_73090545671  136329900_10529659391_316129   
343140    2443347  465784886_73090545671  136329900_10529659391_316129   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
343136      confirm  2017-03-31 15:15:46  
343137       step_3  2017-03-31 15:14:53  
343138       step_2  2017-03-31 15:12:08  
343139       step_1  2017-03-31 15:11:37  
343140        start  2017-03-31 15:10:47  

[330368 rows x 5 columns]

- Keeping duplicate rows as they capture multiple process steps within the same session.
- These entries are essential for reconstructing the full user journey chronologically.

In [79]:
df_data2 = pd.read_csv("initial_datasets/df_final_web_data_pt_2.txt")
df_data2.head()

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58

In [80]:
print(df_data2.shape)
print(df_data2.dtypes)
print(f'\nnull values \n{df_data2.isnull().sum()}')
print(f'\nduplicated values\n{df_data2.duplicated().sum()}')
print(f'\nduplicated rows\n{df_data2[df_data2.duplicated()]}')

(412264, 5)
client_id        int64
visitor_id      object
visit_id        object
process_step    object
date_time       object
dtype: object

null values 
client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

duplicated values
8669

duplicated rows
        client_id             visitor_id                      visit_id  \
125       3751049  937925261_74821699083  628817270_45049432060_475841   
143       9912156  746623513_52101087617  780044921_57678038107_984942   
326       5786602  893139880_61416949586  747153834_69434320996_858907   
328       5786602  893139880_61416949586  747153834_69434320996_858907   
375       5786602  893139880_61416949586   119968876_5423296886_390511   
...           ...                    ...                           ...   
412120    1672289  658797407_91003097342  430110748_90347503221_597628   
412163    4408245   14058946_73585431797  994148957_84373994761_991865   
412165    4408245   14058946_735854

In [81]:
df_data2[df_data2.duplicated(subset=['visitor_id', 'visit_id'], keep=False)]

client_id             visitor_id                      visit_id  \
1         6019349  442094451_91531546617  154620534_35331068705_522317   
2         6019349  442094451_91531546617  154620534_35331068705_522317   
3         6019349  442094451_91531546617  154620534_35331068705_522317   
4         6019349  442094451_91531546617  154620534_35331068705_522317   
5         6019349  442094451_91531546617  154620534_35331068705_522317   
...           ...                    ...                           ...   
412258    9668240   388766751_9038881013   922267647_3096648104_968866   
412259    9668240   388766751_9038881013   922267647_3096648104_968866   
412260    9668240   388766751_9038881013   922267647_3096648104_968866   
412261    9668240   388766751_9038881013   922267647_3096648104_968866   
412262    9668240   388766751_9038881013   922267647_3096648104_968866   

       process_step            date_time  
1           confirm  2017-06-01 11:59:27  
2            step_3  2017-06-01 11:58:48  
3            step_2  2017-06-01 11:58:08  
4            step_1  2017-06-01 11:57:58  
5             start  2017-06-01 11:57:54  
...             ...                  ...  
412258       step_1  2017-05-24 18:48:02  
412259        start  2017-05-24 18:46:10  
412260        start  2017-05-24 18:45:29  
412261       step_1  2017-05-24 18:44:51  
412262        start  2017-05-24 18:44:34  

[402553 rows x 5 columns]

In [82]:
df_final_exp_client = pd.read_csv("initial_datasets/df_final_experiment_clients.txt")
df_final_exp_client.head()

client_id Variation
0    9988021      Test
1    8320017      Test
2    4033851   Control
3    1982004      Test
4    9294070   Control

In [83]:
print(df_final_exp_client.shape)
print(df_final_exp_client.dtypes)
print(f'\nnull values \n{df_final_exp_client.isnull().sum()}')
print(f'\nduplicated values\n{df_final_exp_client.duplicated().sum()}')
print(f'\nunique value \n{df_final_exp_client['Variation'].unique()}')
print(f'\nduplicated values\n{df_final_exp_client.duplicated().sum()}')

(70609, 2)
client_id     int64
Variation    object
dtype: object

null values 
client_id        0
Variation    20109
dtype: int64

duplicated values
0

unique value 
['Test' 'Control' nan]

duplicated values
0


In [84]:
df_web_data= pd.concat([df_data1, df_data2], ignore_index=True)
df_web_data

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

In [85]:
df_web_data.columns
df_web_data.isnull().sum()

client_id       0
visitor_id      0
visit_id        0
process_step    0
date_time       0
dtype: int64

In [86]:
df_exp_demo = df_final_exp_client.merge(df_final_demo, on="client_id", how="inner")
df_exp_demo

client_id Variation  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0        9988021      Test             5.0              64.0      79.0     U   
1        8320017      Test            22.0             274.0      34.5     M   
2        4033851   Control            12.0             149.0      63.5     M   
3        1982004      Test             6.0              80.0      44.5     U   
4        9294070   Control             5.0              70.0      29.0     U   
...          ...       ...             ...               ...       ...   ...   
70604    2443347       NaN            17.0             204.0      52.5     M   
70605    8788427       NaN            10.0             121.0      59.5     M   
70606     266828       NaN            14.0             172.0      60.5     M   
70607    1266421       NaN            13.0             164.0      33.0     M   
70608    9895983       NaN            10.0             121.0      49.0     M   

       num_accts         bal  calls_6_mnth  logons_6_mnth  
0            2.0   189023.86           1.0            4.0  
1            2.0    36001.90           5.0            8.0  
2            2.0   142642.26           5.0            8.0  
3            2.0    30231.76           1.0            4.0  
4            2.0    34254.54           0.0            3.0  
...          ...         ...           ...            ...  
70604        3.0   164693.48           7.0            7.0  
70605        2.0   203149.13           3.0            3.0  
70606        2.0  1135068.68           7.0            7.0  
70607        3.0    27701.79           2.0            3.0  
70608        2.0    97748.22           2.0            2.0  

[70609 rows x 10 columns]

In [87]:
df_exp_demo.isnull().sum()

client_id               0
Variation           20109
clnt_tenure_yr         14
clnt_tenure_mnth       14
clnt_age               15
gendr                  14
num_accts              14
bal                    14
calls_6_mnth           14
logons_6_mnth          14
dtype: int64

In [88]:
df_exp_demo=df_exp_demo.dropna()
df_exp_demo.isnull().sum()

client_id           0
Variation           0
clnt_tenure_yr      0
clnt_tenure_mnth    0
clnt_age            0
gendr               0
num_accts           0
bal                 0
calls_6_mnth        0
logons_6_mnth       0
dtype: int64

In [89]:
df_full = df_exp_demo.merge(df_web_data, on="client_id", how="inner")
df_full

client_id Variation  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  \
0         9988021      Test             5.0              64.0      79.0     U   
1         9988021      Test             5.0              64.0      79.0     U   
2         9988021      Test             5.0              64.0      79.0     U   
3         9988021      Test             5.0              64.0      79.0     U   
4         9988021      Test             5.0              64.0      79.0     U   
...           ...       ...             ...               ...       ...   ...   
321190    6334360      Test             4.0              53.0      66.5     U   
321191    6334360      Test             4.0              53.0      66.5     U   
321192    6334360      Test             4.0              53.0      66.5     U   
321193    6334360      Test             4.0              53.0      66.5     U   
321194    6334360      Test             4.0              53.0      66.5     U   

        num_accts        bal  calls_6_mnth  logons_6_mnth  \
0             2.0  189023.86           1.0            4.0   
1             2.0  189023.86           1.0            4.0   
2             2.0  189023.86           1.0            4.0   
3             2.0  189023.86           1.0            4.0   
4             2.0  189023.86           1.0            4.0   
...           ...        ...           ...            ...   
321190        2.0   93103.47           6.0            9.0   
321191        2.0   93103.47           6.0            9.0   
321192        2.0   93103.47           6.0            9.0   
321193        2.0   93103.47           6.0            9.0   
321194        2.0   93103.47           6.0            9.0   

                   visitor_id                      visit_id process_step  \
0        580560515_7732621733  781255054_21935453173_531117       step_3   
1        580560515_7732621733  781255054_21935453173_531117       step_2   
2        580560515_7732621733  781255054_21935453173_531117       step_3   
3        580560515_7732621733  781255054_21935453173_531117       step_2   
4        580560515_7732621733  781255054_21935453173_531117       step_3   
...                       ...                           ...          ...   
321190  629124187_65258232847  586047816_14599436811_389185       step_2   
321191  629124187_65258232847  586047816_14599436811_389185       step_1   
321192  629124187_65258232847  586047816_14599436811_389185       step_2   
321193  629124187_65258232847  586047816_14599436811_389185       step_1   
321194  629124187_65258232847  586047816_14599436811_389185        start   

                  date_time  
0       2017-04-17 15:27:07  
1       2017-04-17 15:26:51  
2       2017-04-17 15:19:22  
3       2017-04-17 15:19:13  
4       2017-04-17 15:18:04  
...                     ...  
321190  2017-04-11 08:45:38  
321191  2017-04-11 08:44:43  
321192  2017-04-11 08:44:30  
321193  2017-04-11 08:42:36  
321194  2017-04-11 08:42:20  

[321195 rows x 14 columns]

In [90]:
df_full.isnull().sum()

client_id           0
Variation           0
clnt_tenure_yr      0
clnt_tenure_mnth    0
clnt_age            0
gendr               0
num_accts           0
bal                 0
calls_6_mnth        0
logons_6_mnth       0
visitor_id          0
visit_id            0
process_step        0
date_time           0
dtype: int64

In [91]:
# Create a folder named "data_after_cleaning" if it doesn't already exist
os.makedirs("data_after_cleaning", exist_ok=True)

In [92]:
# Save as CSV file (comma-separated)
df_web_data.to_csv("data_after_cleaning/df_web_data.csv", index=False)
df_exp_demo.to_csv("data_after_cleaning/df_exp_demo.csv", index=False)
df_full.to_csv("data_after_cleaning/df_full.csv", index=False)

In [93]:
print(f'df_web_data : \n{df_web_data.columns}')
print(f'\ndf_exp_demo : \n{df_exp_demo.columns}')

df_web_data : 
Index(['client_id', 'visitor_id', 'visit_id', 'process_step', 'date_time'], dtype='object')

df_exp_demo : 
Index(['client_id', 'Variation', 'clnt_tenure_yr', 'clnt_tenure_mnth',
       'clnt_age', 'gendr', 'num_accts', 'bal', 'calls_6_mnth',
       'logons_6_mnth'],
      dtype='object')


In [94]:
df_web_data['date_time'] = pd.to_datetime(df_web_data['date_time'])
print(df_web_data.dtypes)

client_id                int64
visitor_id              object
visit_id                object
process_step            object
date_time       datetime64[ns]
dtype: object


In [95]:
df_exp_demo.dtypes

client_id             int64
Variation            object
clnt_tenure_yr      float64
clnt_tenure_mnth    float64
clnt_age            float64
gendr                object
num_accts           float64
bal                 float64
calls_6_mnth        float64
logons_6_mnth       float64
dtype: object

In [96]:
print(df_exp_demo.describe())

          client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  5.048700e+04    50487.000000      50487.000000  50487.000000   
mean   5.006180e+06       12.031810        150.416365     47.319795   
std    2.877445e+06        6.860326         81.945403     15.518616   
min    5.550000e+02        2.000000         33.000000     17.000000   
25%    2.515688e+06        6.000000         82.000000     33.500000   
50%    5.025043e+06       11.000000        136.000000     48.000000   
75%    7.477922e+06       16.000000        192.000000     59.500000   
max    9.999832e+06       55.000000        669.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth  
count  50487.000000  5.048700e+04  50487.000000   50487.000000  
mean       2.254580  1.495168e+05      3.093272       6.131856  
std        0.533675  3.020390e+05      2.188009       2.175441  
min        1.000000  2.378944e+04      0.000000       3.000000  
25%        2.000000  3.987718e+04  

- clnt_tenure_yr: Clients have, on average, 12 years with Vanguard

- clnt_tenure_mnth: The range (33 to 669 months), confirms long-term relationships

- Ages span from 17 to 96, with a median around 48

- Most clients have 2 accounts (25%, 50%, and 75% quartiles are all 2), though some go up to 7.

- Big skew here: mean ~149,500 but max = 16.3 million.

- Average is about 3 calls per client, with many making 0–6 calls

- Median and mean are both around 6 logins






# Distribution of Client Ages

In [97]:
fig_age = px.histogram(df_exp_demo, x='clnt_age', nbins=30, title='Age Distribution of Clients')
fig_age.update_layout(xaxis_title='Age', yaxis_title='Count')
fig_age.show()

# Account Balance Distribution

In [98]:
fig_raw = px.histogram(df_exp_demo, x='bal', nbins=50, 
                       title='Raw Account Balance Distribution')
fig_raw.update_layout(xaxis_title='Balance', yaxis_title='Client Count')
fig_raw.show()

The variable bal (account balance) is extremely skewed, meaning a small number of clients have very high balances while most have much lower ones. This distorts histograms and statistical models.

To fix this, we applied a logarithmic transformation using log1p (log(1 + x)), which compresses large values and reveals the distribution's true shape more clearly, making visualizations and insights more meaningful.

In [99]:
df_exp_demo.loc[:, 'log_bal'] = np.log1p(df_exp_demo['bal'])
fig_bal = px.histogram(df_exp_demo, x='log_bal', nbins=30, 
                       title='Log-Scaled Distribution of Account Balance')
fig_bal.update_layout(xaxis_title='Log(Balance + 1)', yaxis_title='Count')
fig_bal.show()

C:\Users\sabri\AppData\Local\Temp\ipykernel_2008\584872990.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



# Number of Calls vs. Logins

In [ ]:
fig_calls_logins = px.scatter(df_exp_demo, x='calls_6_mnth', y='logons_6_mnth',
                              title='Client Engagement: Calls vs Logins',
                              labels={'calls_6_mnth': 'Calls (6 months)', 'logons_6_mnth': 'Logins (6 months)'})
fig_calls_logins.show()

# Tenure in Years

In [ ]:
fig_tenure = px.histogram(df_exp_demo, x='clnt_tenure_yr', nbins=20,
                          title='Client Tenure (in Years)')
fig_tenure.update_layout(xaxis_title='Years with Vanguard', yaxis_title='Number of Clients')
fig_tenure.show()

## Who are the primary clients using this online process?

In [133]:
# we identify clients who used the online process

online_ids = df_web_data['client_id'].dropna().unique()

# we match those clients to demographic information

df_online_clients = df_exp_demo[df_exp_demo['client_id'].isin(online_ids)]

print(df_online_clients.describe())

# We compare with full client base

print(df_exp_demo.describe())

          client_id  clnt_tenure_yr  clnt_tenure_mnth      clnt_age  \
count  5.048700e+04    50487.000000      50487.000000  50487.000000   
mean   5.006180e+06       12.031810        150.416365     47.319795   
std    2.877445e+06        6.860326         81.945403     15.518616   
min    5.550000e+02        2.000000         33.000000     17.000000   
25%    2.515688e+06        6.000000         82.000000     33.500000   
50%    5.025043e+06       11.000000        136.000000     48.000000   
75%    7.477922e+06       16.000000        192.000000     59.500000   
max    9.999832e+06       55.000000        669.000000     96.000000   

          num_accts           bal  calls_6_mnth  logons_6_mnth       log_bal  
count  50487.000000  5.048700e+04  50487.000000   50487.000000  50487.000000  
mean       2.254580  1.495168e+05      3.093272       6.131856     11.320933  
std        0.533675  3.020390e+05      2.188009       2.175441      0.933887  
min        1.000000  2.378944e+04      0.000

- Around 47 years old on average

- Clients for about 12 years, with over half having more than 10 years of tenure

- Holding 2 accounts (most common number)

- Logging in about 6 times over 6 months

- Maintaining moderate account balances (with a median around 65,700)




## Are the primary clients younger or older, new or long-standing?

They are clearly on the older side (most between 34 and 60), and definitely long-standing — the majority have been clients for over a decade.
